numpy/core/src/multiarray/refcount.c

```diff
@@ -51,6 +51,31 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
             PyArray_Item_INCREF(data + offset, new);
         }
     }
+    else if (PyDataType_HASSUBARRAY(descr)) {
+        int size, i, inner_elsize;
+
+        inner_elsize = descr->subarray->base->elsize;
+        /* Subarrays are always contiguous in memory */
+        size = descr->elsize / inner_elsize;
+
+        if (descr->subarray->base->type_num == NPY_OBJECT) {
+            /* Fast path for only-objects */
```

Comment:
This path isn't used in `PyDataType_HASFIELDS` - I don't see why we should optimize for it here either.